src/ReinforcementLearningZoo/src/algorithms/policy_gradient/ppo.jl

```diff
@@ -242,37 +249,39 @@ function _update!(p::PPOPolicy, t::AbstractTrajectory)
         dims = 2,
         terminal = t[:terminal],
     )
-    returns = advantages .+ select_last_dim(states_plus_values, 1:n_rollout)
+    returns = to_device(advantages .+ select_last_dim(states_plus_values, 1:n_rollout))
+    advantages = to_device(advantages)
 
     actions_flatten = flatten_batch(select_last_dim(t[:action], 1:n))
-    action_log_probs = select_last_dim(t[:action_log_prob], 1:n)
+    action_log_probs = select_last_dim(to_device(t[:action_log_prob]), 1:n)
 
     # TODO: normalize advantage
     for epoch in 1:n_epochs
         rand_inds = shuffle!(rng, Vector(1:n_envs*n_rollout))
         for i in 1:n_microbatches
             inds = rand_inds[(i-1)*microbatch_size+1:i*microbatch_size]
             if t isa MaskedPPOTrajectory
-                lam = send_to_device(
-                    D,
-                    select_last_dim(
-                        flatten_batch(select_last_dim(t[:legal_actions_mask], 2:n+1)),
-                        inds,
-                    ),
+                lam = select_last_dim(
+                    flatten_batch(select_last_dim(LAM, 2:n+1)),
+                    inds,
                 )
+                
             else
                 lam = nothing
             end
-            s = send_to_device(D, select_last_dim(states_flatten, inds))  # !!! performance critical
-            a = send_to_device(D, select_last_dim(actions_flatten, inds))
+
+            # s = to_device(select_last_dim(states_flatten_on_host, inds))
+            # !!! we need to convert it into a continuous CuArray otherwise CUDA.jl will complain scalar indexing
+            s = to_device(collect(select_last_dim(states_flatten_on_host, inds)))
+            a = to_device(collect(select_last_dim(actions_flatten, inds)))
```

Comment:
Sometimes it is `select_last_dim(to_device(`, and sometimes it is switched like here. I'm sure there is thought behind it but I have not really understood exactly what is going on with all of them.

What would happen if we here (our maybe outside the loop) did `to_device` first and then did the `select_last_dim`? To me it feels like that is what you have tried to do in other places, but I have probably missed something.

Reply:
Technically it is always better to send data `to_device` first. But here the collect trajectory might be too large sometimes. So I guess it's safe to send each sampled batch to the device lazily. It's hard to balance here because the overall performance depends on the size of trajectory, `n_epochs` and `n_microbatches`.

Reply:
A special note is that the continuous `@view` of a `CuArray` is converted to a `CuArray` directly in the latest version, instead of a `SubArray`. But if it's not continuous, then that might be a problem in the following-up calculations. But if even if a `@view` of `Array` is continuous, when converting to `CuArray`, it will be *adapt*ed and the whole parent of that view is transferred to GPU first, then a `SubArray` of that `CuArray` is returned, instead of a continuous `CuArray`. Maybe we can propose some improvements to `CUDA.jl`.